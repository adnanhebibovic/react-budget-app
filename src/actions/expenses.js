import db from '../db/firebase'
import {v4 as uuid} from 'uuid'

export const addExpenseToStore = function(expense = {}) {
    const defaults = {
        id: new uuid(),
        title: '',
        amount: 0,
        createdAt: new Date().getTime(),
        ...expense
    };

    return {
        type: 'ADD_EXPENSE',
        expense: {
            id: defaults.id,
            title: defaults.title,
            amount: defaults.amount,
            createdAt: defaults.createdAt
        }
    };
}

export const addExpense = function(expense = {}) {
    const defaults = {
        createdAt: new Date().getTime(),
        ...expense
    }

    return function(dispatch) {
        return db.collection('expenses').add(defaults)
        .then((doc) => {
            dispatch(addExpenseToStore({
                id: doc.id,
                ...defaults
            }))
        })
    }
}

export const removeExpenseFromStore = function(id) {
    return {
        type: 'REMOVE_EXPENSE',
        id
    }
}

export const removeExpense = function(id) {
    return function(dispatch) {
        return db.collection('expenses').doc(id).delete()
        .then(() => {
            dispatch(removeExpenseFromStore(id))
        })
    }
}

export const editExpenseInStore = function(id, expense) {
    const defaults = {
        title: '',
        amount: 0,
        ...expense
    };

    return {
        type: 'EDIT_EXPENSE',
        id,
        expense: defaults
    }
}

export const editExpense = function(id, expense) {
    return function(dispatch) {
        return db.collection('expenses').doc(id).update(expense)
        .then(() => {
            dispatch(editExpenseInStore(id, expense))
        })
    }
}