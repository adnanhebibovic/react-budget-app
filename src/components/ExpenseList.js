import React from 'react'
import { connect } from 'react-redux'

import ExpenseListItem from './ExpenseListItem'
import select from '../selectors/expenses'

export function ExpenseList(props) {
    return (
        <div className="content-container">
            <div className="list-header">
                <div className="show-for-mobile">Expenses</div>
                <div className="show-for-desktop">Expense</div>
                <div className="show-for-desktop">Amount</div>
            </div>
            {props.expenses.length == 0 && <div className="list-item list-item--empty"><span>Please create an expense to get started!</span></div>}
            {props.expenses.map((expense) => {
                return <ExpenseListItem key={expense.id} {...expense}></ExpenseListItem>
            })}
        </div>
    )
}

const mapStateToProps = (state) => {
    return {
        expenses: select(state.expenses, state.filters)
    }
}

export default connect(mapStateToProps)(ExpenseList)
