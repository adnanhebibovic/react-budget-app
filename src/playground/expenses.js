import { createStore, combineReducers } from 'redux'
import uuid from 'uuid'

const addExpense = function({title = '', amount = 0} = {}) {
    return {
        type: 'ADD_EXPENSE',
        expense: {
            id: uuid(),
            title,
            amount,
            createdAt: new Date()
        }
    }
}

const removeExpense = function(id = uuid()) {
    return {
        type: 'REMOVE_EXPENSE',
        id
    }
}

const editExpense = function(id = uuid(), expense = {}) {
    return {
        type: 'EDIT_EXPENSE',
        id,
        expense
    }
}

const expensesDefaultState = []

const expensesReducer = function(state = expensesDefaultState, action) {
    switch (action.type) {
        case 'ADD_EXPENSE':
            return state.concat(action.expense)
        case 'REMOVE_EXPENSE':
            return state.filter((expense) => expense.id !== action.id)
        case 'EDIT_EXPENSE':
            return state.map((expense) => expense.id == action.id 
                ? {...expense, ...action.expense}
                : expense
            )
        default:
            return state;
    }
}

const setTextFilter = function(text = undefined) {
    return {
        type: 'SET_TEXT_FILTER',
        text
    }
}

const sortByAmount = function() {
    return {
        type: 'SORT_BY',
        sortBy: 'amount'
    }
}

const sortByDate = function() {
    return {
        type: 'SORT_BY',
        sortBy: 'createdAt'
    }
}

const setDateFilters = function({startDate = undefined, endDate = undefined} = {}) {
    return {
        type: 'SET_DATE_FILTERS',
        dates: {
            startDate,
            endDate
        }
    }
}

const filtersDefaultState = {
    text: undefined,
    sortBy: undefined,
    startDate: undefined,
    endDate: undefined
}

const filtersReducer = function(state = filtersDefaultState, action) {
    switch (action.type) {
        case 'SET_TEXT_FILTER':
            return {
                ...state,
                text: action.text
            }
        case 'SORT_BY':
            return {
                ...state,
                sortBy: action.sortBy
            }
        case 'SET_DATE_FILTERS':
            return {
                ...state,
                ...action.dates
            }
        default:
            return state;
    }
}

const getExpenses = function(expenses, {text, sortBy, startDate, endDate}) {
    const result = expenses.filter((expense) => {
        const textMatch = typeof text === 'undefined' || expense.title.includes(text);
        const startDateMatch = typeof startDate === 'undefined' || expense.createdAt >= startDate;
        const endDateMatch = typeof endDate === 'undefined' || expense.createdAt <= endDate;

        return textMatch && startDateMatch && endDateMatch;
    })

    if (sortBy)
    {
        return result.sort((e1, e2) => {
            if (e1[sortBy] < e2[sortBy]) {
                return -1;
            } else if (e1[sortBy] > e2[sortBy]) {
                return 1;
            } else {
                return 0;
            }
        })
    }

    return result
}

const store = createStore(
    combineReducers({
        expenses: expensesReducer,
        filters: filtersReducer
    })
)

store.dispatch(addExpense({
    title: 'January bills',
    amount: 279.00 
}))

store.dispatch(addExpense({
    title: 'February bills',
    amount: 523.00 
}))

store.dispatch(addExpense({
    title: 'March bills',
    amount: 358.00 
}))

store.dispatch(addExpense({
    title: 'New car',
    amount: 1000.00 
}))

store.subscribe(() => {
    const state = store.getState();
    console.log(getExpenses(state.expenses, state.filters))
})

store.dispatch(setTextFilter('bills'))
store.dispatch(sortByAmount())
store.dispatch(sortByDate())
store.dispatch(setDateFilters({endDate: new Date()}))
