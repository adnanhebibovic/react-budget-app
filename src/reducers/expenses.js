export const expensesDefaultState = []

export default function(state = expensesDefaultState, action) {
    switch (action.type) {
        case 'ADD_EXPENSE':
            return state.concat(action.expense)
        case 'SET_EXPENSE':
            return state.concat(action.expenses.filter((expense) => state.findIndex((item) => item.id === expense.id) < 0))
        case 'REMOVE_EXPENSE':
            return state.filter((expense) => expense.id !== action.id)
        case 'EDIT_EXPENSE':
            return state.map((expense) => expense.id === action.id 
                ? {...expense, ...action.expense}
                : expense
            )
        default:
            return state;
    }
}