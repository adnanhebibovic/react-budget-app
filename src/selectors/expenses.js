export default function(expenses, {text, sortBy, startDate, endDate}) {
    const result = expenses.filter((expense) => {
        const textMatch = typeof text == 'undefined' || expense.title.toLowerCase().includes(text.toLowerCase());
        const startDateMatch = typeof startDate == 'undefined' || expense.createdAt > startDate;
        const endDateMatch = typeof endDate == 'undefined' || expense.createdAt < endDate;

        return textMatch && startDateMatch && endDateMatch;
    })

    if (sortBy) {
        return result.sort((e1, e2) => {
            if (e1[sortBy] < e2[sortBy]) {
                return -1;
            } else if (e1[sortBy] > e2[sortBy]) {
                return 1;
            } else {
                return 0;
            }
        })
    }

    return result;
}
