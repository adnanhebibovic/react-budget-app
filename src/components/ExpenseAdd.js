import React from 'react'
import { connect } from 'react-redux'
import { useHistory } from 'react-router-dom'
import { startAddExpense } from '../actions/expenses'

import ExpenseForm from './ExpenseForm'

function ExpenseAdd(props) {
    let history = useHistory()

    function onSubmit(expense) {
        props.addExpense(props.uid, expense);
        history.push('/');
    } 

    return (
        <div>
           <h1>Create Expense</h1>
            <ExpenseForm onSubmit={onSubmit}></ExpenseForm>
        </div>
    )
}

const mapStateToProps = (state) => {
    return {
        expenses: state.expenses,
        uid: state.auth.uid
    }
}

const mapDispatchToProps = (dispatch) => {
    return {
        addExpense: (uid, expense) => dispatch(startAddExpense(uid, expense))
    }
}

export default connect(mapStateToProps, mapDispatchToProps)(ExpenseAdd)
