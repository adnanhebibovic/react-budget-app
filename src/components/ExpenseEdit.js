import React from 'react'
import { connect } from 'react-redux'
import { useParams, useHistory } from 'react-router-dom'
import { startEditExpense, startRemoveExpense } from '../actions/expenses'

import ExpenseForm from './ExpenseForm'

function ExpenseEdit(props) {
    let { id } = useParams()
    let history = useHistory()

    function onSubmit(expense) {
        props.editExpense(id, expense)
        history.push('/')
    }

    function onRemove() {
        props.removeExpense(id)
        history.push('/')
    }

    return (
        <div>
            <h1>Edit Expense</h1>
            <ExpenseForm expense={props.expenses.find((e) => e.id === id)} onSubmit={onSubmit}/>
            <button onClick={onRemove}>Remove</button>
        </div>
    )
}

const mapStateToProps = (state, props) => {
    return {
        expenses: state.expenses
    }
}

const mapDispatchToProps = (dispatch, props) => {
    return {
        editExpense: (id, expense) => dispatch(startEditExpense(id, expense)),
        removeExpense: (id) => dispatch(startRemoveExpense(id))
    }
}

export default connect(mapStateToProps, mapDispatchToProps)(ExpenseEdit)
