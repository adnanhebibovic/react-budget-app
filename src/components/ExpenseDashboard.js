import React from 'react';
import ExpenseList from './ExpenseList'
import ExpenseListFilters from './ExpenseListFilters'
import ExpenseSummary from './ExpenseSummary';

export default function ExpenseDashboard() {
    return (
        <div>
            <h1>Expense dashboard</h1>
            <ExpenseListFilters></ExpenseListFilters>
            <ExpenseSummary></ExpenseSummary>
            <ExpenseList></ExpenseList>
        </div>
    );
}