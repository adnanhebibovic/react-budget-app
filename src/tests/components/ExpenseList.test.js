import React from 'react'
import { BrowserRouter } from 'react-router-dom'
import { create, act } from 'react-test-renderer';

import { ExpenseList } from '../../components/ExpenseList'
import expenses from '../fixtures/expenses'

test('Should render ExpenseList with expenses', () => {
    let root
    act(() => {
        root = create(<BrowserRouter><ExpenseList expenses={expenses}></ExpenseList></BrowserRouter>)
    });
    expect(root).toMatchSnapshot()
})

test('Should render ExpenseList with no expenses', () => {
    let root
    act(() => {
        root = create(<BrowserRouter><ExpenseList expenses={[]}></ExpenseList></BrowserRouter>)
    });
    expect(root).toMatchSnapshot()
})