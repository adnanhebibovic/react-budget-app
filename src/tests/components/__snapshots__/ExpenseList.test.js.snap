// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render ExpenseList with expenses 1`] = `
<div>
  <h2>
    Expense list
  </h2>
  <div>
    <a
      aria-current={null}
      href="/expenses/1"
      onClick={[Function]}
    >
      <h3>
        first expense
      </h3>
    </a>
    <p>
      400.0
    </p>
    <time
      dateTime={"2015-03-24T23:00:00.000Z"}
    >
      Wed Mar 25 2015 00:00:00 GMT+0100
    </time>
  </div>
  <div>
    <a
      aria-current={null}
      href="/expenses/2"
      onClick={[Function]}
    >
      <h3>
        second expense
      </h3>
    </a>
    <p>
      500.0
    </p>
    <time
      dateTime={"2015-03-27T23:00:00.000Z"}
    >
      Sat Mar 28 2015 00:00:00 GMT+0100
    </time>
  </div>
  <div>
    <a
      aria-current={null}
      href="/expenses/3"
      onClick={[Function]}
    >
      <h3>
        third expense
      </h3>
    </a>
    <p>
      200.0
    </p>
    <time
      dateTime={"2015-03-26T23:00:00.000Z"}
    >
      Fri Mar 27 2015 00:00:00 GMT+0100
    </time>
  </div>
  <div>
    <a
      aria-current={null}
      href="/expenses/4"
      onClick={[Function]}
    >
      <h3>
        fourth expense
      </h3>
    </a>
    <p>
      100.0
    </p>
    <time
      dateTime={"2015-03-25T23:00:00.000Z"}
    >
      Thu Mar 26 2015 00:00:00 GMT+0100
    </time>
  </div>
</div>
`;

exports[`Should render ExpenseList with no expenses 1`] = `
<div>
  <h2>
    Expense list
  </h2>
  <p>
    Please create an expense to get started!
  </p>
</div>
`;
